torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+            # are underway in the post_backward stream. See:
+            # github.com/NVIDIA/apex/blob/master/apex/parallel/distributed.py
+            orig_grad_data.record_stream(self._streams["post_backward"])
+
+    def _queue_wait_for_post_backward(self) -> None:
+        """Try to queue a `wait_for_post_backward` callback.
+        Only called on root and only queue one callback at the beginning of
+        outer most backward.
+        """
+        assert self._is_root
+        if not self._post_backward_callback_queued:
+            self._assert_state([TrainingState.IDLE])
+            self._post_backward_callback_queued = True
+            Variable._execution_engine.queue_callback(self._wait_for_post_backward)
+
+    @torch.no_grad()
+    def _wait_for_post_backward(self) -> None:
+        """Wait for post-backward to finish. Only called on root instance."""
+        assert self._is_root
+        # Check if the root module has params and if any of them has
+        # the `requires_grad` field set. If `requires_grad=False` for
+        # all the params, the post_backward hook will not fire and the
+        # state will remain in `TrainingState.BACKWARD_PRE`.
+        if any([p.requires_grad for p in self.params]):
+            self._assert_state(TrainingState.BACKWARD_POST)
+        else:
+            self._assert_state(TrainingState.BACKWARD_PRE)
+
+        torch.cuda.current_stream().wait_stream(self._streams["post_backward"])
+        # A backward pass is done, clean up below.
+
+        def _remove_shard_bwd_hook(fsdp_module: FullyShardedDataParallel) -> None:
+            """Helper used below on all fsdp modules."""
+            for p in fsdp_module.params:
+                if p.requires_grad:
+                    if hasattr(p, "_shard_bwd_hook"):
+                        assert len(p._shard_bwd_hook) == 2, len(p._shard_bwd_hook)  # type: ignore[attr-defined]
+                        p._shard_bwd_hook[1].remove()  # type: ignore[attr-defined]
+                        delattr(p, "_shard_bwd_hook")
+
+        # Update root and nested FSDP's hooks and flags.
+        for m in self.modules():  # includes self
+            if isinstance(m, FullyShardedDataParallel):
+                _remove_shard_bwd_hook(m)
+                m._pre_backward_hook_has_run = False
+                if any(p.requires_grad for p in m.parameters()):
+                    # Check if the module has params and if any of them has
+                    # the `requires_grad` field set. If `requires_grad=False` for
+                    # all the params, the post_backward hook will not fire and the
+                    # state will remain in `TrainingState.BACKWARD_PRE`.
+                    if any([p.requires_grad for p in m.params]):
+                        m._assert_state(TrainingState.BACKWARD_POST)
+                    else:
+                        m._assert_state(TrainingState.BACKWARD_PRE)
```

Comment:
duplicate code and comments with the following lines? Shall we consolidate?

https://github.com/pytorch/pytorch/blob/1604f3e5a86c93113fb2e14eabf9bcf98b832a1c/torch/distributed/_fsdp/fully_sharded_data_parallel.py#L618-L625

Reply:
the cases are actually different here. The last case is to check cases when all the submodules of the FSDP wrapped module do not require grads. The first two cases are checking for parameters only in this FSDP wrapped module